File: src/report.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { DOMWidgetModel, DOMWidgetView, ISerializers } from '@jupyter-widgets/base';

import {
  Report,
  models,
  VisualDescriptor,
  Page,
  IReportEmbedConfiguration,
} from 'powerbi-client';

import { MODULE_NAME, MODULE_VERSION } from './version';

// Import the CSS
import '../css/report.css';
import { getActivePageSize, getRequestedPage, powerbi, setTokenExpirationListener, getTokenExpirationTimeout } from './utils';

const EXPORT_DATA_DEFAULT_STATE: ExportVisualDataRequest = {
  pageName: undefined,
  visualName: undefined,
  rows: undefined,
  exportDataType: undefined,
};

const REPORT_FILTER_REQUEST_DEFAULT_STATE = {
  filters: [],
  request_completed: true,
};

const REPORT_NOT_EMBEDDED_MESSAGE = 'Power BI report is not embedded';

export class ReportModel extends DOMWidgetModel {
  defaults(): any {
    return {
      ...super.defaults(),
      _model_name: ReportModel.model_name,
      _model_module: ReportModel.model_module,
      _model_module_version: ReportModel.model_module_version,
      _view_name: ReportModel.view_name,
      _view_module: ReportModel.view_module,
      _view_module_version: ReportModel.view_module_version,
      _embed_config: {},
      _embedded: false,
      container_height: 0,
      container_width: 0,
      _export_visual_data_request: EXPORT_DATA_DEFAULT_STATE,
      _visual_data: null,
      _event_data: {
        event_name: null,
        event_details: null,
      },
      _get_filters_request: false,
      _report_filters: [],
      _report_filters_request: REPORT_FILTER_REQUEST_DEFAULT_STATE,
      _get_pages_request: false,
      _report_pages: [],
      _get_visuals_page_name: null,
      _page_visuals: [],
      _report_bookmark_name: null,
      _get_bookmarks_request: false,
      _report_bookmarks: [],
      _report_active_page: null,
      _token_expired: false,
      _client_error: null,
      _init_error: null
    };
  }

  static serializers: ISerializers = {
    ...DOMWidgetModel.serializers,
  };

  static model_name = 'ReportModel';
  static model_module = MODULE_NAME;
  static model_module_version = MODULE_VERSION;
  static view_name = 'ReportView';
  static view_module = MODULE_NAME;
  static view_module_version = MODULE_VERSION;
}

interface ExportVisualDataRequest {
  pageName?: string;
  visualName?: string;
  rows?: number;
  exportDataType?: number;
}

interface ReportFilterRequest {
  filters: models.ReportLevelFilters[];
  request_completed: boolean;
}

interface DOMRectSize {
  bottom: number;
  height: number;
  left: number;
  right: number;
  top: number;
  width: number;
  x: number;
  y: number;
}

export class ReportView extends DOMWidgetView {
  report: Report;
  reportContainer: HTMLDivElement;

  render(): void {
    const newDivElement = document.createElement('div');
    newDivElement.style.visibility = 'hidden';
    this.el.appendChild(newDivElement);

    this.reportContainer = newDivElement;

    this.embedConfigChanged();

    // Observe changes in the traitlets in Python, and define custom callback.
    this.model.on('change:_embed_config', this.embedConfigChanged, this);
    this.model.on('change:container_height', this.containerSizeChanged, this);
    this.model.on('change:container_width', this.containerSizeChanged, this);
    this.model.on('change:_export_visual_data_request', this.exportVisualDataRequestChanged, this);
    this.model.on('change:_get_filters_request', this.getFiltersRequestChanged, this);
    this.model.on('change:_report_filters_request', this.reportFiltersChanged, this);
    this.model.on('change:_get_pages_request', this.getPagesRequestChanged, this);
    this.model.on('change:_get_visuals_page_name', this.getVisualsPageNameChanged, this);
    this.model.on('change:_report_bookmark_name', this.reportBookmarkNameChanged, this);
    this.model.on('change:_get_bookmarks_request', this.getBookmarksRequestChanged, this);
    this.model.on('change:_report_active_page', this.reportActivePageChanged, this);
  }

  containerSizeChanged(): void {
    if (this.reportContainer) {
      this.reportContainer.style.height = `${this.model.get('container_height')}px`;
      this.reportContainer.style.width = `${this.model.get('container_width')}px`;
    }
  }

  setTokenExpiredFlag(): void {
    this.model.set('_token_expired', true);
    this.touch();
  }

  embedConfigChanged(): void {
    const embedConfig = this.model.get('_embed_config');
    const reportConfig = embedConfig as IReportEmbedConfiguration;

    if (this.report) {
      const accessToken = reportConfig.accessToken as string;

      // To avoid re-embedding the report
      if (
        this.report.config.embedUrl === reportConfig.embedUrl &&
        this.report.config.accessToken !== reportConfig.accessToken
      ) {
        // Set new access token
        this.report.setAccessToken(accessToken);

        if (reportConfig.accessToken) {
          // Set token expiration listener to update the token TOKEN_REFRESH_THRESHOLD minutes before expiration
          setTokenExpirationListener(reportConfig.accessToken, this);
        }
      }
      this.model.set('_embedded', true);
      this.touch();
      return;
    } else {
      // Refresh access token before embedding if token is expired
      if (!reportConfig.accessToken || getTokenExpirationTimeout(reportConfig.accessToken) <= 0) {
        this.setTokenExpiredFlag();
        return;
      }
    }

    // Flag for checking create mode
    const createReportMode =
      embedConfig.viewMode !== models.ViewMode.View &&
      embedConfig.viewMode !== models.ViewMode.Edit;

    // Check if the embedding mode is create mode
    if (createReportMode) {
      // Create blank Power BI report
      this.report = powerbi.createReport(this.reportContainer, reportConfig as models.IReportCreateConfiguration) as Report;
    } else {
      // Phased loading
      this.report = powerbi.load(this.reportContainer, reportConfig) as Report;
    }

    // Set default aspect ratio
    let aspectRatio = 9 / 16;

    this.report.on('loaded', async () => {
      console.log('Loaded');

      if (reportConfig.accessToken) {
        // Set token expiration listener to update the token TOKEN_REFRESH_THRESHOLD minutes before expiration
        setTokenExpirationListener(embedConfig.accessToken, this);
      }

      try {
        // Check if embed mode is view/edit to get active page size
        if (!createReportMode) {
          // Get page size for the Power BI report
          const { width, height } = await getActivePageSize(this.report);
          if (width && height) {
            // Update aspect ratio according to the page size
            aspectRatio = height / width;
          } else {
            console.error('Invalid report size');
          }
        }

        let width = this.model.get('container_width') as number;
        let height = this.model.get('container_height') as number;

        if (!width || !height) {
          // Get dimensions of output cell
          const DOMRect: DOMRectSize = this.el.getBoundingClientRect();

          width = DOMRect.width || 980;
          height = width * aspectRatio;
        }

        // Set dimensions of report container
        this.reportContainer.style.width = `${width}px`;
        this.reportContainer.style.height = `${height}px`;

        
        // Show the report container
        this.reportContainer.style.visibility = 'visible';

        if (!createReportMode) {
          // Complete the phased embedding
          this.report.render();
        }
      } catch (error) {
        console.error(error);
      }

      // Invoke loaded event handler on kernel side
      this.model.set('_event_data', {
        event_name: 'loaded',
        event_details: null,
      });

      this.touch();
    });

    this.report.on('rendered', () => {
      console.log('Rendered');
      // Invoke rendered event handler on kernel side
      this.model.set('_event_data', {
        event_name: 'rendered',
        event_details: null,
      });

      this.touch();
    });

    this.report.on('error', (errorMessage) => {
      // Invoke error handling on kernel side if container isn't visible yet
      if (this.reportContainer.style.visibility === "visible") {
        return;
      }

      const messageDetail = (errorMessage as any)?.detail;
      this.model.set('_init_error', `${messageDetail?.message} - ${messageDetail?.detailedMessage}`);
      this.touch();
    });

    // Notify that report embedding has started
    this.model.set('_embedded', true);

    // Commit model changes
    this.touch();
  }

  async exportVisualDataRequestChanged(): Promise<void> {
    if (!this.report) {
      this.logError(REPORT_NOT_EMBEDDED_MESSAGE);
      return;
    }

    const exportVisualDataRequest = this.model.get(
      '_export_visual_data_request'
    ) as ExportVisualDataRequest;

    // Check export visual data request object is null or empty
    if (
      !exportVisualDataRequest.pageName &&
      !exportVisualDataRequest.visualName &&
      !exportVisualDataRequest.rows &&
      !exportVisualDataRequest.exportDataType
    ) {
      // This is the case of model reset
      return;
    }

    if (!exportVisualDataRequest.pageName || !exportVisualDataRequest.visualName) {
      const errorMessage = 'Page and visual names are required';
      this.logError(errorMessage);

      return;
    }

    const pageName = exportVisualDataRequest.pageName;
    const visualName = exportVisualDataRequest.visualName;
    const dataRows = exportVisualDataRequest.rows;
    const exportDataType = exportVisualDataRequest.exportDataType;

    try {
      const selectedPage: Page = await getRequestedPage(this.report, pageName);
      const visuals: VisualDescriptor[] = await selectedPage.getVisuals();
      const selectedVisual: VisualDescriptor = visuals.filter((visual: VisualDescriptor) => {
        return visual.name === visualName;
      })[0];

      if (!selectedVisual) {
        throw 'Visual not found';
      }

      const data = await selectedVisual.exportData(exportDataType, dataRows);

      // Update data
      this.model.set('_visual_data', data.data);
      this.touch();
    } catch (error) {
      this.logError(error);
    }
  }

  async getFiltersRequestChanged(): Promise<void> {
    if (!this.report) {
      this.logError(REPORT_NOT_EMBEDDED_MESSAGE);
      return;
    }

    const get_filters_request = this.model.get('_get_filters_request') as boolean;
    if (!get_filters_request) {
      // Reset of get filters request
      return;
    }

    try {
      // Get list of filters applied on the report
      const filters: models.IFilter[] = await this.report.getFilters();

      if (!filters) {
        throw 'No filters available';
      }

      this.model.set('_report_filters', filters);
      this.model.set('_get_filters_request', false);
      this.touch();
    } catch (error) {
      this.logError(error);
    }
  }

  async reportFiltersChanged(): Promise<void> {
    if (!this.report) {
      this.logError(REPORT_NOT_EMBEDDED_MESSAGE);
      return;
    }

    const filterRequest = this.model.get('_report_filters_request') as ReportFilterRequest;

    if (filterRequest.request_completed) {
      // Reset of filter request
      return;
    }

    try {
      // Add new filters or remove filters when filters array is empty
      if (filterRequest.filters.length > 0) {
        await this.report.updateFilters(models.FiltersOperations.Replace, filterRequest.filters);
      } else {
        await this.report.updateFilters(models.FiltersOperations.RemoveAll);
      }

      // Reset filter request
      this.model.set('_report_filters_request', REPORT_FILTER_REQUEST_DEFAULT_STATE);
      this.touch();
    } catch (error) {
      this.logError(error);
    }
  }

  async getPagesRequestChanged(): Promise<void> {
    if (!this.report) {
      this.logError(REPORT_NOT_EMBEDDED_MESSAGE);
      return;
    }

    const get_pages_request = this.model.get('_get_pages_request') as boolean;
    if (!get_pages_request) {
      return;
    }

    try {
      const pages: Page[] = await this.report.getPages();

      if (!pages) {
        throw 'Pages not found';
      }

      // Remove 'report' property from Page object to handle nested property loop
      const pagesWithoutReport = pages.map((page) => {
        const { report, ...newPage } = page;
        return newPage;
      });

      this.model.set('_report_pages', pagesWithoutReport);
      this.touch();
    } catch (error) {
      this.logError(error);
    }
  }

  async getVisualsPageNameChanged(): Promise<void> {
    if (!this.report) {
      this.logError(REPORT_NOT_EMBEDDED_MESSAGE);
      return;
    }

    const get_visuals_page_name = this.model.get('_get_visuals_page_name') as string;
    if (!get_visuals_page_name) {
      return;
    }

    try {
      const selectedPage: Page = await getRequestedPage(this.report, get_visuals_page_name);
      const visuals: VisualDescriptor[] = await selectedPage.getVisuals();

      if (!visuals) {
        throw 'Visuals not found';
      }

      // Remove 'page' property from Visual object to handle nested property loop
      const visualsWithoutPage = visuals.map((visual) => {
        const { page, ...newVisual } = visual;
        return newVisual;
      });

      this.model.set('_page_visuals', visualsWithoutPage);
      this.touch();
    } catch (error) {
      this.logError(error);
    }
  }

  async reportBookmarkNameChanged(): Promise<void> {
    if (!this.report) {
      this.logError(REPORT_NOT_EMBEDDED_MESSAGE);
      return;
    }

    const bookmarkName = this.model.get('_report_bookmark_name') as string;

    try {
      // Apply corresponding bookmark to the embedded report
      await this.report.bookmarksManager.apply(bookmarkName);
    } catch (error) {
      this.logError(error);
    }
  }

  async getBookmarksRequestChanged(): Promise<void> {
    if (!this.report) {
      this.logError(REPORT_NOT_EMBEDDED_MESSAGE);
      return;
    }

    const get_bookmarks_request = this.model.get('_get_bookmarks_request') as boolean;
    if (!get_bookmarks_request) {
      return;
    }

    try {
      // Get list of bookmarks present in the report
      const bookmarks: models.IReportBookmark[] = await this.report.bookmarksManager.getBookmarks();

      // Check if there is any bookmark saved on the embedded report
      if (bookmarks.length === 0) {
        this.model.set('_report_bookmarks', ['']);
        this.touch();
      } else {
        this.model.set('_report_bookmarks', bookmarks);
        this.touch();
      }
    } catch (error) {
      this.logError(error);
    }
  }

  async reportActivePageChanged(): Promise<void> {
    if (!this.report) {
      this.logError(REPORT_NOT_EMBEDDED_MESSAGE);
      return;
    }

    const pageName = this.model.get('_report_active_page') as string;

    try {
      // Set the provided page as active
      await this.report.setPage(pageName);
    } catch (error) {
      this.logError(error);
    }
  }

  private logError(errorMessage: any): void {
    let stringifiedError = JSON.stringify(errorMessage);
    if (stringifiedError === '{}') {
      stringifiedError = errorMessage.toString();
    }

    console.error(errorMessage);
    this.model.set('_client_error', stringifiedError);
    this.touch();
  }
}

File: powerbi_client/report.py
#!/usr/bin/env python
# coding: utf-8

# Copyright (c) Microsoft.

"""
TODO: Add module docstring
"""

import time

from IPython import get_ipython
from ipywidgets import DOMWidget
from jupyter_ui_poll import ui_events
from traitlets import Bool, Dict, Float, Unicode, List, observe

from ._frontend import module_name, module_version


class Report(DOMWidget):
    """PowerBI report embedding widget"""

    # Name of the widget view class in front-end
    _view_name = Unicode('ReportView').tag(sync=True)

    # Name of the widget model class in front-end
    _model_name = Unicode('ReportModel').tag(sync=True)

    # Name of the front-end module containing widget view
    _view_module = Unicode('powerbi-client-frontend').tag(sync=True)

    # Name of the front-end module containing widget model
    _model_module = Unicode('powerbi-client-frontend').tag(sync=True)

    # Version of the front-end module containing widget view
    _view_module_version = Unicode('^0.1.0').tag(sync=True)

    # Version of the front-end module containing widget model
    _model_module_version = Unicode('^0.1.0').tag(sync=True)

    # Default values for widget traits
    VISUAL_DATA_DEFAULT_STATE = ''
    EXPORT_VISUAL_DATA_REQUEST_DEFAULT_STATE = {}
    REGISTERED_EVENT_HANDLERS_DEFAULT_STATE = {}
    EVENT_DATA_DEFAULT_STATE = {
        'event_name': None,
        'event_details': None
    }
    REPORT_FILTER_REQUEST_DEFAULT_STATE = {
        'filters': [],
        'request_completed': True
    }
    GET_PAGES_REQUEST_DEFAULT_STATE = False
    REPORT_PAGES_DEFAULT_STATE = []
    GET_VISUALS_DEFAULT_PAGE_NAME = ''
    PAGE_VISUALS_DEFAULT_STATE = []

    # Other constants
    REPORT_NOT_EMBEDDED_MESSAGE = "Power BI report is not embedded"

    # Process upto n UI events per iteration
    PROCESS_EVENTS_ITERATION = 3

    # Check for UI every n seconds
    POLLING_INTERVAL = 0.5

    # Widget specific properties.
    # Widget properties are defined as traitlets. Any property tagged with `sync=True`
    # is automatically synced to the frontend *any* time it changes in Python.
    # It is synced back to Python from the frontend *any* time the model is touched in frontend.

    # TODO: Add trait validation
    embed_config = Dict(None).tag(sync=True)
    _embedded = Bool(False).tag(sync=True)

    container_height = Float(0).tag(sync=True)
    container_width = Float(0).tag(sync=True)

    # TODO: Add trait validation
    # TODO: Start with _
    export_visual_data_request = Dict(None).tag(sync=True)
    # TODO: Start with _
    visual_data = Unicode(VISUAL_DATA_DEFAULT_STATE).tag(sync=True)

    _event_data = Dict(EVENT_DATA_DEFAULT_STATE).tag(sync=True)

    # TODO: Add trait validation
    _report_filters_request = Dict(REPORT_FILTER_REQUEST_DEFAULT_STATE).tag(sync=True)

    _get_pages_request = Bool(GET_PAGES_REQUEST_DEFAULT_STATE).tag(sync=True)
    _report_pages = List(REPORT_PAGES_DEFAULT_STATE).tag(sync=True)

    _get_visuals_page_name = Unicode(GET_VISUALS_DEFAULT_PAGE_NAME).tag(sync=True)
    _page_visuals = List(PAGE_VISUALS_DEFAULT_STATE).tag(sync=True)

    # Methods
    def __init__(self, access_token, embed_url, token_type=0, **kwargs):
        # Tells if Power BI events are being observed
        self._observing_events = False

        # Registered Power BI event handlers methods
        self._registered_event_handlers = dict(
            self.REGISTERED_EVENT_HANDLERS_DEFAULT_STATE)

        self.set_embed_config(access_token, embed_url, token_type)

        # Init parent class DOMWidget
        super(Report, self).__init__(**kwargs)

    def set_embed_config(self, access_token, embed_url, token_type=0):
        """
        TODO: Add docstring
        """
        self.embed_config = {
            'type': 'report',
            'accessToken': access_token,
            'embedUrl': embed_url,
            'tokenType': token_type
        }
        self._embedded = False

    def set_dimensions(self, container_height, container_width):
        """Set width and height of Power BI report in px

        Args:
            container_height (number): report height
            container_width (number): report width
        """
        self.container_height = container_height
        self.container_width = container_width

    def export_visual_data(self, page_name, visual_name, rows=10, underlying_data=False):
        """Returns the data of given visual of the embedded Power BI report

        Args:
            page_name (string): Page name of the embedded report
            visual_name (string): Visual's unique name 
            rows (int, optional): Number of rows of data. Defaults to 10
            underlying_data (boolean, optional): Choice to show the underlying data or not. Default is False.

        Returns:
            string: visual's exported data
        """
        if self._embedded == False:
            raise Exception(self.REPORT_NOT_EMBEDDED_MESSAGE)

        # Start exporting data on client side
        self.export_visual_data_request = {
            'pageName': page_name,
            'visualName': visual_name,
            'rows': rows,
            'underlyingData': underlying_data
        }

        # Check if ipython kernel is available
        if get_ipython():
            # Wait for client-side to send visual data
            with ui_events() as ui_poll:
                # While visual data is not received
                while self.visual_data == self.VISUAL_DATA_DEFAULT_STATE:
                    ui_poll(self.PROCESS_EVENTS_ITERATION)
                    time.sleep(self.POLLING_INTERVAL)

        exported_data = self.visual_data

        # Reset the export_visual_data_request and visual_data's value
        self.export_visual_data_request = dict(self.EXPORT_VISUAL_DATA_REQUEST_DEFAULT_STATE)
        self.visual_data = self.VISUAL_DATA_DEFAULT_STATE

        return exported_data

    def on(self, event, callback):
        """Register a callback to execute when the report emits the target event
        Parameters

        Args:
            event (string): Name of Power BI event (eg. 'loaded', 'rendered', 'error')
            callback (function): User defined function. Callback function is invoked with event details as parameter
        """
        # TODO: Value of event should be from one of the Report.allowedEvents array
        self._registered_event_handlers[event] = callback

        def get_event_data(change):

            event_data = change.new
            event_name = event_data['event_name']
            event_details = event_data['event_details']

            # Do not invoke callback when _event_data trait is reset
            if event_name is None:
                return

            # Check if a handler is registered for the current event
            if event_name not in self._registered_event_handlers:
                return

            event_handler = self._registered_event_handlers[event_name]
            event_handler(event_details)

            # Reset the _event_data trait, so as to receive next event
            self._event_data = dict(self.EVENT_DATA_DEFAULT_STATE)

        # Check if already observing events
        if not self._observing_events:

            # Prevents calling DOMWidget.observe() again
            self._observing_events = True

            # Start observing Power BI events
            self.observe(get_event_data, '_event_data')

    def update_filters(self, filters):
        """Set report level filters in the embedded report.
            Currently supports models.FiltersOperations.Add

        Args:
            filters ([models.ReportLevelFilters]): List of report level filters

        Raises:
            Exception: When report is not embedded
        """
        if self._embedded == False:
            raise Exception(self.REPORT_NOT_EMBEDDED_MESSAGE)

        self._report_filters_request = {
            'filters': filters,
            'request_completed': False
        }

        # TODO: Should we wait for the filters to be applied

    def remove_filters(self):
        """Remove all report level filters from the embedded report

        Raises:
            Exception: When report is not embedded
        """
        self.update_filters([])

    def get_pages(self):
        """Returns the list of pages of the embedded Power BI report

        Returns:
            string: list of pages
        """
        if self._embedded == False:
            raise Exception(self.REPORT_NOT_EMBEDDED_MESSAGE)

        # Start getting pages on client side
        self._get_pages_request = True

        # Check if ipython kernel is available
        if get_ipython():
            # Wait for client-side to send list of pages
            with ui_events() as ui_poll:
                # While list of report pages is not received
                while self._report_pages == self.REPORT_PAGES_DEFAULT_STATE:
                    ui_poll(self.PROCESS_EVENTS_ITERATION)
                    time.sleep(self.POLLING_INTERVAL)

        pages = self._report_pages

        # Reset the get_pages_request and report_pages's value
        self._get_pages_request = bool(self.GET_PAGES_REQUEST_DEFAULT_STATE)
        self._report_pages = list(self.REPORT_PAGES_DEFAULT_STATE)

        return pages

    def visuals_on_page(self, page_name):
        """Returns the list of visuals of the given page of the embedded Power BI report

        Args:
            page_name (string): Page name of the embedded report

        Returns:
            string: list of visuals
        """
        if self._embedded == False:
            raise Exception(self.REPORT_NOT_EMBEDDED_MESSAGE)

        # Start getting visuals on client side
        self._get_visuals_page_name = page_name

        # Check if ipython kernel is available
        if get_ipython():
            # Wait for client-side to send list of visuals
            with ui_events() as ui_poll:
                # While list of visuals is not received
                while self._page_visuals == self.PAGE_VISUALS_DEFAULT_STATE:
                    ui_poll(self.PROCESS_EVENTS_ITERATION)
                    time.sleep(self.POLLING_INTERVAL)

        visuals = self._page_visuals

        # Reset the get_visuals_page_name and page_visuals's value
        self._get_visuals_page_name = self.GET_VISUALS_DEFAULT_PAGE_NAME
        self._page_visuals = list(self.PAGE_VISUALS_DEFAULT_STATE)

        return visuals
